File: 프로그래머스/lv2/12978. 배달/배달.py
import heapq

def dikjstra(n, graph, k) -> list:
    count = [int(1e9)] * (n + 1)
    heap = [[0, 1]] # 최단 거리, 현재 위치
    count[1] = 0
    while heap:
        cnt, pos = heapq.heappop(heap)
        if cnt > count[pos] or cnt > k:
            continue
        for node in graph[pos]:
            next_pos, next_cnt = node[0], node[1] + cnt
            if count[next_pos] > next_cnt and next_cnt <= k:
                count[next_pos] = next_cnt
                heapq.heappush(heap, [next_cnt, next_pos])
    return count
    

def solution(N, road, K):
    graph = [[] for _ in range(N + 1)]
    for info in road:
        sour, des, cnt = info
        graph[sour].append([des, cnt])
        graph[des].append([sour, cnt])
    
    count = dikjstra(N, graph, K)
    answer = 0
    for data in count:
        if int(1e9) != data:
            answer += 1
    return answer

File: 프로그래머스/lv2/12978. 배달/배달.java
import java.util.*;

class Solution {
    private List<int[]>[] graph;
    public int solution(int N, int[][] road, int K) {
        int answer = 0;
        graph = new ArrayList[N + 1];
        for(int i=0 ; i<=N ; i++){
            graph[i] = new ArrayList<>();
        }
        
        for(int[] r : road){
            int sour = r[0], des = r[1], cost = r[2];
            graph[sour].add(new int[]{des, cost});
            graph[des].add(new int[]{sour, cost});
        }
        answer = dikjstra(N, K);
        return answer;
    }
    
    private int dikjstra(int N, int k){
        Queue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);
        int[] costs = new int[N + 1];
        Arrays.fill(costs, Integer.MAX_VALUE);
        pq.add(new int[]{1, 0});
        costs[1] = 0;
        
        while(!pq.isEmpty()){
            int[] cur = pq.poll();
            int sour = cur[0], cnt = cur[1];
            if(cnt > costs[sour])continue;
            for(int[] next : graph[sour]){
                int des = next[0], nextCost = next[1] + costs[sour];
                if(costs[des] > nextCost){
                    costs[des] = nextCost;
                    pq.add(new int[]{des, nextCost});
                }
            }
        }
        
        int ret = 0;
        for(int i = 1 ; i<=N ; i++){
            if(k >= costs[i])ret++;
        }
        return ret;
    }
}